### ISSUE TITLE
Installing link-grammar with python bindings in anaconda python environment

### ISSUE THREAD

**Issue by christian-storm** [opencog/link-grammar#298]
After many hours trying to make this work I finally figured it out.  I'm posting the recipe here so that others won't have to endure my pain.

  Installation in anaconda
 ===================
My environment (link_grammar) was created by 
%conda create --name link_grammar python=3.5

To fix the problem:

> ImportError: No module named 'linkgrammar'

I had to create a file named linkgrammar.pth in the directory /Users/xxxx/anaconda/envs/link_grammar/lib/python3.5/site-packages

The linkgrammar.pth file has one line in it:
linkgrammar/

See /Users/xxxx/anaconda/envs/link_grammar/lib/python3.5/site.py for details on appending module search paths for third-party packages to sys.path.

To fix the error:

> ImportError: dlopen(/Users/xxxx/anaconda/envs/link_grammar/lib/python3.5/site-packages/linkgrammar/_clinkgrammar.so, 2): Library not loaded: libpython3.5m.dylib
>   Referenced from: /Users/xxxx/anaconda/envs/link_grammar/lib/python3.5/site-packages/linkgrammar/_clinkgrammar.so
>   Reason: image not found

referred to [here](https://github.com/ContinuumIO/anaconda-issues/issues/498) I had to add an @rpath to libpython3.5m.dylib:

```
%cd /Users/xxxx/anaconda/lib
%otool -L libpython3.5m.dylib #No rpath listed
%install_name_tool -id @rpath/libpython3.5m.dylib libpython3.5m.dylib
%otool -L libpython3.5m.dylib #Should see rpath now
```

To make sure the proper installation directory was being targeted I had to add a prefix to the configure command:

```
%./configure --disable-java-bindings --enable-python-bindings --enable-editline --prefix=/Users/xxxx/anaconda/envs/link_grammar/
%make
%sudo make install
```

**Comment by linas** [comment]
Thanks!

Ideally, the makefiles should be modified to perform all of these steps automatically, when building on MacOS.  However, making these changes is probably quite tedious, as it requires a lot of knowledge of the automake/autoconf system.

Another possibility is to put this into the README file, mostly a verbitim cut-n-paste of above...

**Comment by linas** [comment]
There might be partial fixes in the next releste ... also, the next release will have `--enable-python3-bindings` in it.

**Comment by linas** [comment]
I don't understand the rpath issues on OSX.  On Linux, the build process does this:
```
/bin/bash ../../libtool  --tag=CXX   --mode=link g++  -g -O2 -O3 -Wall -std=c++11 -version-info 8:8:3 -L/usr/lib -lpython3.4m -module  -o _clinkgrammar.la -rpath /usr/local/lib/python3.4/site-packages/linkgrammar ../../bindings/python3/_clinkgrammar_la-lg_python_wrap.lo  ../../link-grammar/liblink-grammar.la -laspell
```

so python3.4m is linked in (it should be 3.5m for you)

**Comment by ampli** [comment]
I don't understand the issue too.
`libpython3.5m.dylib` is not created by the link-grammar build process, so it shouldn't modify its rpath.

However, this rpath problem doesn't exist with Python3.5 version installed via the HomeBrew installer (Installed on an up-to-date El-Capitan). (See below for this Python version details.)

But this version has another problem: The linking fails due to an error in the `PYTHON_EXTRA_LDFLAGS`, which incorrectly includes `-u _PyMacError`, a symbol that was once defined in the library but it is no more defined there. After removing it in `lib/python3.5/config-3.5m/Makefile` (from which the configuration script indirectly gets it), the bindings are created just fine and the library passes the installcheck tests.

I reported this problem to HomeBrew (by opening a GitHub issue).

**Also**:
I tried to remove **linkgrammar.pth** from the system (after the installation).
It turns out that `import linkgrammar` succeeds just fine even without it:
```
Python 3.5.2 (default, Aug 16 2016, 05:35:40) 
[GCC 4.2.1 Compatible Apple LLVM 7.3.0 (clang-703.0.31)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import linkgrammar
>>> dir(linkgrammar)
['Clinkgrammar', 'Dictionary', 'Link', 'Linkage', 'ParseOptions', 'Sentence', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', '__version__', '_clinkgrammar', 'clinkgrammar', 'linkgrammar']
```

I couldn't understand why it was needed, unless the tested Python version was buggy, or an incorrect PYTHONPATH was used. So most probably it can be eventually removed.

BTW,  "make check" is not working due to another problem, related to a needed rpath in the LG library so the pre-installed library will be found. **This I would like to fix**. (In previous OS X releases DYLD_LIBRARY_PATH could be used for that, but on El-Capitan its use is limited due to a new system security feature called [SIP](https://en.wikipedia.org/wiki/System_Integrity_Protection).)

**Comment by linas** [comment]
Closing, stale.  I think this was fixed in version 5.3.10 which included a `linkgrammar.pth` that gets installed at `.../wherever/python3.5/site-packages`

**Comment by jmugan** [comment]
I'm having similar problems with 5.4.2. I have Anaconda. When I do 

`./configure --disable-java-bindings --enable-python-bindings=3 --enable-editline --prefix=/anaconda3/`

I get
```
link-grammar-5.4.2  build configuration settings

	prefix:                         /usr/local
	C compiler:                     gcc  -DUSE_SAT_SOLVER=1 -D_DEFAULT_SOURCE -g -O2 -O3
	C++ compiler:                   g++  -DUSE_SAT_SOLVER=1 -D_DEFAULT_SOURCE -std=c++11 -stdlib=libc++ -g -O2 -O3 -Wall
	Error handler per-thread:       yes
	Editline command-line history:  no
	UTF8 editline support:          no
	Java libraries:                 no
	Java interfaces:                
	Swig interfaces generator:      yes
	Perl interfaces:                no
	Perl install location:          
	Python2 interfaces:             no
	Python3 interfaces:             3.5
	ASpell spell checker:           no
	HunSpell spell checker:         no
	HunSpell dictionary location:   
	Boolean SAT parser:             yes (using the bundled minisat library)
	SQLite-backed dictionary:       no
	Viterbi algorithm parser:       no
	Corpus statistics database:     no
	RegEx tokenizer:                no
```

then
`make`

I get

```Making all in python3
  CXX      _clinkgrammar_la-lg_python_wrap.lo
rm: .libs/_clinkgrammar_la-lg_python_wrap.o: Permission denied
error: unable to open output file '.libs/_clinkgrammar_la-lg_python_wrap.o': 'Operation not permitted'
1 error generated.
make[2]: *** [_clinkgrammar_la-lg_python_wrap.lo] Error 1
make[1]: *** [all-recursive] Error 1
make: *** [all-recursive] Error 1
```

Any ideas?

**Comment by ampli** [comment]
Can it be that you once tries to run `make` under root?
Here is how to check the problem:
```
make clean
make --print-directory 
```

Then check the permissions of the relevant directory that is shown in the directory trace:
`ls -la /FULL_PATH/.libs`

If the owner is not you, or you don't have write permission there for some reason, try to fix the permission of the whole LG source directory (owner and/or write permissions, as needed):
```
# chown -R youruser LGDIR
$ chmod -R 755 LGDIR
```
---
Anyway, I tried to recompile and install LG 5.4.2 with Anaconda, and it works fine (despite of many compilation warnings that got fixed by now). . It passes all the tests (which means it works fine with Python - I validated that the only python I have at that point is python3 from of Anaconda). My Anaconda installation is under my home directory (`~anaconda`) but I guess it shouldn't make a difference.

BTW:
1) In the same occasion I also tries it with Homebrew's current python3 (3.6.2) and it also passes all the tests.
2) I also tried the current development version (rc 4.5.4?). Besides new clang warnings (due to changes done from 4.5.2), it has a severe locale problem (it doesn't pass tests with non-ASCII characters), and its invocation hints on a stack garbage problem. I will fix that and send a PR.
3) Because a system may have several python versions installed, it could be a good idea to modify `configure` to show (in the configuration setup summary) which one got used. This can be done by
displaying the binary path (because `configure` actually uses the first python in the path).
4) It would also be nice to have an API to get the configuration options of the library.

**Comment by jmugan** [comment]
Sorry, I deleted my comment because I realized I needed `sudo` and felt stupid.

But now when I `import linkgrammar` I get `ImportError: dlopen(/anaconda3/lib/python3.5/site-packages/linkgrammar/_clinkgrammar.so, 2): Library not loaded: libpython3.5m.dylib`

If it fix it by `export DYLD_LIBRARY_PATH=/anaconda3/lib` I get
`Fatal Python error: PyThreadState_Get: no current thread
Abort trap: 6`

**Comment by ampli** [comment]
> I realized I needed sudo
[
But note that it still a better idea to always compile under your user instead of under root.
You just need to install the sources as your user (not using root). 

> Fatal Python error: PyThreadState_Get: no current thread Abort trap: 6

It looks like you have more than one python3 version in your system, and the one `configure` has picked up is different than the one you tried to use.

For more hints on this possible problem, you can compare the `configure` output to the location of the include files according to python3:
```
$ ./configure ... > /tmp/configure.out
$ grep 'Python include path' /tmp/configure.out
[Output: Location of Python's include files]

$ python3
import sysconfig
sysconfig.get_paths()
[Validate, according to the include file location, if this is the correct python3.]
```

To see all the python3 in your system:
```
$ which python3                                       # the location of the python3 you use
$ sudo which python3                              # maybe its PATH is different than in your user
$ sudo /usr/libexec/locate.updatedb        # your current directory should not be `/`
$ locate /python3 | grep 'python3$'
[All locations of python3.]
```

**Comment by ampli** [comment]
In continuation to my previous post here: 
You can use `which -a python3 ` to find all python3 in your path, and, if need, change the PATH components order so the desired python3 will be first and then run `configure` again.

**Comment by jmugan** [comment]
Thanks for your help!
`grep 'Python include path' /tmp/configure.out`
gives
`checking for Python include path... -I/anaconda3/include/python3.5m`

`sysconfig.get_paths()`
gives
`{'stdlib': '/anaconda3/lib/python3.5', 'platinclude': '/anaconda3/include/python3.5m', 'platstdlib': '/anaconda3/lib/python3.5', 'platlib': '/anaconda3/lib/python3.5/site-packages', 'data': '/anaconda3', 'scripts': '/anaconda3/bin', 'purelib': '/anaconda3/lib/python3.5/site-packages', 'include': '/anaconda3/include/python3.5m'}`

`which python3`
gives
`/anaconda3/bin/python3`

`sudo which python3`
gives
`/anaconda3/bin/python3`

`locate /python3 | grep 'python3$'`
gives
```
/Users/jmugan/bin/link-grammar-5.4.2/bindings/python3
/Users/jmugan/google-cloud-sdk/.install/.backup/platform/gsutil/third_party/crcmod/python3
/Users/jmugan/google-cloud-sdk/.install/.backup/platform/gsutil/third_party/httplib2/python3
/Users/jmugan/google-cloud-sdk/platform/gsutil/third_party/crcmod/python3
/Users/jmugan/google-cloud-sdk/platform/gsutil/third_party/httplib2/python3
/anaconda3/bin/python3
/anaconda3/lib/python3.5/site-packages/astroid/tests/testdata/python3
/anaconda3/pkgs/astroid-1.5.3-py35hc14dc17_0/lib/python3.5/site-packages/astroid/tests/testdata/python3
/anaconda3/pkgs/astroid-1.5.3-py36h1333018_0/lib/python3.6/site-packages/astroid/tests/testdata/python3
/anaconda3/pkgs/ipykernel-4.6.1-py35h9865fc8_0/share/jupyter/kernels/python3
/anaconda3/pkgs/ipykernel-4.6.1-py36h3208c25_0/share/jupyter/kernels/python3
/anaconda3/pkgs/python-3.5.4-h8f450c2_22/bin/python3
/anaconda3/pkgs/python-3.6.3-h6804ab2_0/bin/python3
/anaconda3/share/jupyter/kernels/python3
/usr/local/share/jupyter/kernels/python3
```

and `which -a python3`
gives
`/anaconda3/bin/python3`

**Comment by jmugan** [comment]
I did all of the clean you suggested to make me the owner of the directories. I do `make` and everything looks fine, but when I do `make check` I get 
```============================================================================
See bindings/python-examples/test-suite.log
Please report to link-grammar@googlegroups.com
============================================================================
```
And the log says
```=================================================================
   link-grammar 5.4.2: bindings/python-examples/test-suite.log
=================================================================

# TOTAL: 1
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: tests
===========

Fatal Python error: PyThreadState_Get: no current thread
```

**Comment by linas** [comment]
Well, actually, I think you'd want to search for multiple copies of `libpython3` 

A full stack trace would be the most useful it would tell us what is throwing the exception.   I'm assuming that this is sufficient to capture it:

https://stackoverflow.com/questions/17859132/how-to-get-exception-detail-in-python3

Perhaps there is some easier way, I don't know.

**Comment by linas** [comment]
The interwebs tell me that this would also be useful:  `otool -L /anaconda3/lib/python3.5/site-packages/linkgrammar/_clinkgrammar.so`  this should print all of the shared libraries that got linked.  They should not have any surprises.

The `no current thread` is bizarre.  Another source of information would be to strace the thing to see what is happening.  Google tells me that the equivalent of strace on OSX is either `dtruss` or `dtrace`

**Comment by jmugan** [comment]
`locate libpython3`
gives
```
/Library/Frameworks/SAVI.framework/Versions/A/Frameworks/Python.framework/Versions/3.5/lib/libpython3.5.dylib
/Library/Frameworks/SAVI.framework/Versions/A/Frameworks/Python.framework/Versions/3.5/lib/libpython3.5m.dylib
/Library/Frameworks/SAVI.framework/Versions/A/Frameworks/Python.framework/Versions/3.5/lib/python3.5/config-3.5m/libpython3.5.a
/Library/Frameworks/SAVI.framework/Versions/A/Frameworks/Python.framework/Versions/3.5/lib/python3.5/config-3.5m/libpython3.5.dylib
/Library/Frameworks/SAVI.framework/Versions/A/Frameworks/Python.framework/Versions/3.5/lib/python3.5/config-3.5m/libpython3.5m.a
/Library/Frameworks/SAVI.framework/Versions/A/Frameworks/Python.framework/Versions/3.5/lib/python3.5/config-3.5m/libpython3.5m.dylib
/anaconda3/lib/libpython3.5m.a
/anaconda3/lib/libpython3.5m.dylib
/anaconda3/lib/python3.5/config-3.5m/libpython3.5m.a
/anaconda3/pkgs/python-3.5.4-h8f450c2_22/lib/libpython3.5m.a
/anaconda3/pkgs/python-3.5.4-h8f450c2_22/lib/libpython3.5m.dylib
/anaconda3/pkgs/python-3.5.4-h8f450c2_22/lib/python3.5/config-3.5m/libpython3.5m.a
/anaconda3/pkgs/python-3.6.3-h6804ab2_0/lib/libpython3.6m.a
/anaconda3/pkgs/python-3.6.3-h6804ab2_0/lib/libpython3.6m.dylib
/anaconda3/pkgs/python-3.6.3-h6804ab2_0/lib/python3.6/config-3.6m-darwin/libpython3.6m.a
```
and `otool -L /anaconda3/lib/python3.5/site-packages/linkgrammar/_clinkgrammar.so`
gives
```
/anaconda3/lib/python3.5/site-packages/linkgrammar/_clinkgrammar.so:
	@rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	/anaconda3/lib/liblink-grammar.5.dylib (compatibility version 10.0.0, current version 10.2.0)
	@rpath/libpython3.5m.dylib (compatibility version 3.5.0, current version 3.5.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
```
It won't let me catch the exception, I just get
```
Fatal Python error: PyThreadState_Get: no current thread
Abort trap: 6
```
when I try to `import linkgrammar`

**Comment by ampli** [comment]
> /anaconda3/lib/libpython3.5m.a
> /anaconda3/lib/libpython3.5m.dylib
> /anaconda3/lib/python3.5/config-3.5m/libpython3.5m.a
> /anaconda3/pkgs/python-3.5.4-h8f450c2_22/lib/libpython3.5m.a
> /anaconda3/pkgs/python-3.5.4-h8f450c2_22/lib/libpython3.5m.dylib
> /anaconda3/pkgs/python-3.5.4-h8f450c2_22/lib/python3.5/config-3.5m/libpython3.5m.a
> /anaconda3/pkgs/python-3.6.3-h6804ab2_0/lib/libpython3.6m.a
> /anaconda3/pkgs/python-3.6.3-h6804ab2_0/lib/libpython3.6m.dylib
> /anaconda3/pkgs/python-3.6.3-h6804ab2_0/lib/python3.6/config-3.6m-darwin/libpython3.6m.a

I see that you have 4 different python versions installed (unless the version with no `m` are actually identical - you may use `cmp file1 file2` to find out). At my anaconda installation (which works fine with LG 5.4.2) only libpython3.5 is found.

As a test, I would suggest to eliminate all of them but one (maybe leave only 3.6m). (I guess there is a way to tell `conda` or the like to erase unwanted versions). Then reconfigure and recompile.
(Note that you have to run `locate.updatedb` after you eliminate all the versions but one, if you would like to check that by using `locate`).

**Comment by jmugan** [comment]
Cool. I think the thing I will try next is reinstalling Anaconda. It'll take me a few days to build up the gumption since I'll have to reinstall all the other stuff it uses, but I'll let you know how it turns out. Thanks for your help!

**Comment by linas** [comment]
Why isn't anaconda preventing this mess? This smells of the bad old days of DLL Hell from microsoft windows, how can it be 2017 and there still be a mess?

**Comment by jmugan** [comment]
I changed the path to the old version of Anaconda (to make sure it wasn't being used) and installed the latest version of Anaconda to my local directory, and I still get the same error when I install link-grammar and then try `import linkgrammar`
```
Fatal Python error: PyThreadState_Get: no current thread
Abort trap: 6
```
About a year ago, I had link-grammar installed and working fine using Brew Python. I deleted that Python and started using Anaconda recently. So, somewhere, deep inside my machine, there must be a link that directs something to the wrong place.

**Comment by linas** [comment]
I'm thinking that the `dtreuss` or the `dtrace` would be the most informative. 

But first the exception.   I don't think you actually tried to catch the exception.  I can't beleive that python would crash without catching it.  The code you need is this:
```
import sys, traceback
try:
    import linkgrammar
except:
    traceback.print_exc()
    print ("and now for the stack\n")
    traceback.print_stack()
```
https://docs.python.org/3/library/traceback.html

**Comment by linas** [comment]
For dtruss, Here: https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man1/dtruss.1m.html

so it would be
```
dtruss python
(bunch of stuff is printed, then the > prompt)
import linkgrammar
(more stuff is printed)
```
 The stuff both before and after  is interesting. saying `dtruss -l` is probably a very good idea.    You may need to say `dtruss -a` or `dtruss -s` also. Not sure.

 I'm guessing that dtrace works in a similar way.

**Comment by jmugan** [comment]
I'll try the dtruss. Here's the output of trying to catch the exception.
```
JMugan:bin jmugan$ more test.py
import sys, traceback
try:
    import linkgrammar
except:
    traceback.print_exc()
    print ("and now for the stack\n")
    traceback.print_stack()
JMugan:bin jmugan$ python test.py
Fatal Python error: PyThreadState_Get: no current thread

Abort trap: 6
JMugan:bin jmugan$
```

**Comment by jmugan** [comment]
`sudo dtruss python` (it requires sudo) was giving me a bunch of errors like 
```dtrace: error on enabled probe ID 2165 (ID 947: syscall::write_nocancel:return): invalid kernel access in action #12 at DIF offset 92```
and not much else.

`sudo dtrace python` (also required sudo) said
```
dtrace: system integrity protection is on, some features will not be available

dtrace: no probes specified
```

**Comment by ampli** [comment]
@jmugan,
See [this](http://blog.shalman.org/enabling-dtrace-on-macos-sierra)  for how to enable dtrace.
Also, locate `_clinkgrammar.so`. This is the module that is loaded when `linkgrammar` is imported.
You should only have one such file for python3.
In any case (disregarding if you have more than one) , invoke `python3` with `-v` and then import linkgrammar:
```
$ python3 -v
>>> import linkgrammar
```
It will show you from where it loads `_clinkgrammar.so`.
You can then use `otool -L` on the binary of this python3 (as shown by `which python3`) and on the loaded `_clinkgrammar.so` to see if they use the same libraries.

@linas,
I also rechecked the python detection and configuration code in `configure.ac`, and I see that I don't like it. It is based on traditional autoconf code to configure python, and it looks rather problematic.
Especially, it may have problems with non-system python installations.
I will try to clean it up and to make clear exactly which python it configures.

**Comment by jmugan** [comment]
`python -v` with `import linkgrammar` gives
```>>> import linkgrammar
# /Users/jmugan/anaconda3/lib/python3.6/site-packages/linkgrammar/__pycache__/__init__.cpython-36.pyc matches /Users/jmugan/anaconda3/lib/python3.6/site-packages/linkgrammar/__init__.py
# code object from '/Users/jmugan/anaconda3/lib/python3.6/site-packages/linkgrammar/__pycache__/__init__.cpython-36.pyc'
# /Users/jmugan/anaconda3/lib/python3.6/site-packages/linkgrammar/__pycache__/linkgrammar.cpython-36.pyc matches /Users/jmugan/anaconda3/lib/python3.6/site-packages/linkgrammar/linkgrammar.py
# code object from '/Users/jmugan/anaconda3/lib/python3.6/site-packages/linkgrammar/__pycache__/linkgrammar.cpython-36.pyc'
# /Users/jmugan/anaconda3/lib/python3.6/site-packages/linkgrammar/__pycache__/clinkgrammar.cpython-36.pyc matches /Users/jmugan/anaconda3/lib/python3.6/site-packages/linkgrammar/clinkgrammar.py
# code object from '/Users/jmugan/anaconda3/lib/python3.6/site-packages/linkgrammar/__pycache__/clinkgrammar.cpython-36.pyc'
Fatal Python error: PyThreadState_Get: no current thread

Abort trap: 6
```

`locate` gave me this in addition to the one we want, which looks promising, since it is on the system python.
```
/usr/local/lib/python2.7/site-packages/linkgrammar/_clinkgrammar.so
/usr/local/lib/python2.7/site-packages/pylinkgrammar/_clinkgrammar.so
/usr/local/lib/python3.5/site-packages/linkgrammar/_clinkgrammar.so
```
But getting rid of those packages (deleting the directories) didn't help.

The method for enabling dtrace seems a bit extreme. I worry it could affect something down the road.

**Comment by jmugan** [comment]
Using `otool` I get
```
JMugan:~ jmugan$ otool -L /Users/jmugan/anaconda3/bin/python
/Users/jmugan/anaconda3/bin/python:
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 853.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)
```
and
```
JMugan:~ jmugan$ otool -L /Users/jmugan/anaconda3/lib/python3.6/site-packages/linkgrammar/_clinkgrammar.so
/Users/jmugan/anaconda3/lib/python3.6/site-packages/linkgrammar/_clinkgrammar.so:
	@rpath/libc++.1.dylib (compatibility version 1.0.0, current version 1.0.0)
	/Users/jmugan/anaconda3/lib/liblink-grammar.5.dylib (compatibility version 10.0.0, current version 10.2.0)
	@rpath/libpython3.6m.dylib (compatibility version 3.6.0, current version 3.6.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.0.0)
```

**Comment by linas** [comment]
This is infuriating. How can there be no current thread? Why would a thread exit?   

Do you have access to `gdb`?  The backtrace there would be interesting.  So: run `gdb python` then hit `c` to "continue", then `import linkgrammar` -- presumably, this will enter gdb, and then the stack trace can be printed with `bt`.  There are also ways of examining all threads, and it should be printing every time a thread is created or destroyed. ...

**Comment by jmugan** [comment]
It wouldn't let me use gdb. It said it wasn't cosigned.

**Comment by ampli** [comment]
It is relatively easy to cosign it. However, it may not help because it has not yet been ported to work with shared libraries on Sierra (which has a new shared library format). But you can use the native debugger **lldb**.

**Comment by jmugan** [comment]
I followed a couple of tutorials, but I couldn't get `lldb` to work. I got errors like
```
(lldb) file test.py
error: '/Users/jmugan/bin/test.py' doesn't contain any 'host' platform architectures: x86_64h, x86_64, i386
```
I'll think on if there is any other way to go about figuring out what the problem is.

**Comment by ampli** [comment]
I don't think a stack trace would help in your case, but see below how to do that.  I still guess it is a mismatch of python versions.
(BTW, I don't know if we currently can explicitly find the python version inside `_clinkgrammar*.so`,
but maybe we can deliberately put this info there so in case of similar problems we will know exactly which python got linked with it.).

For this demonstration, I linked against python 2.7.14, and it works:
```
[mac:src/link-grammar-devel/master] amir% python2
Python 2.7.14 (default, Sep 25 2017, 09:53:22) 
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.37)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import linkgrammar
>>>
```
I then tried it in python 2.7.11, and it failed with the infamous "no current thread" that we get on version mismatches:
```
[mac:src/link-grammar-devel/master] amir% env PYTHONPATH=/usr/local/lib/python2.7/site-packages /Users/amir/.pyenv/versions/2.7.11/bin/python
Python 2.7.11 (default, Dec  4 2017, 19:47:31) 
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.38)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import linkgrammar
Fatal Python error: PyThreadState_Get: no current thread
Abort
Exit 134
```

Here is how to get a stack trace using `lldb` (the `env` below is because I use `csh`):
```
[mac:src/link-grammar-devel/master] amir% env PYTHONPATH=/usr/local/lib/python2.7/site-packages lldb /Users/amir/.pyenv/versions/2.7.11/bin/python
(lldb) target create "/Users/amir/.pyenv/versions/2.7.11/bin/python"
Current executable set to '/Users/amir/.pyenv/versions/2.7.11/bin/python' (x86_64).

(lldb) r
Process 81966 launched: '/Users/amir/.pyenv/versions/2.7.11/bin/python' (x86_64)
Python 2.7.11 (default, Dec  4 2017, 19:47:31) 
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.38)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import linkgrammar
Fatal Python error: PyThreadState_Get: no current thread
Process 81966 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
    frame #0: 0x00007fff8c0bfd42 libsystem_kernel.dylib`__pthread_kill + 10
libsystem_kernel.dylib`__pthread_kill:
->  0x7fff8c0bfd42 <+10>: jae    0x7fff8c0bfd4c            ; <+20>
    0x7fff8c0bfd44 <+12>: movq   %rax, %rdi
    0x7fff8c0bfd47 <+15>: jmp    0x7fff8c0b8caf            ; cerror_nocancel
    0x7fff8c0bfd4c <+20>: retq   
Target 0: (python) stopped.

(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x00007fff8c0bfd42 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff8c1ad457 libsystem_pthread.dylib`pthread_kill + 90
    frame #2: 0x00007fff8c025420 libsystem_c.dylib`abort + 129
    frame #3: 0x0000000101a1f395 Python`Py_FatalError + 49
    frame #4: 0x0000000101a1e903 Python`PyThreadState_Get + 28
...
    frame #51: 0x00000001000d8d55 python`PyRun_InteractiveOneFlags + 373
    frame #52: 0x00000001000d879e python`PyRun_InteractiveLoopFlags + 94
    frame #53: 0x00000001000d8660 python`PyRun_AnyFileExFlags + 64
    frame #54: 0x00000001000eeb99 python`Py_Main + 3065
    frame #55: 0x00007fff8bf91235 libdyld.dylib`start + 1
    frame #56: 0x00007fff8bf91235 libdyld.dylib`start + 1
(lldb) 
```

I still suggest that you use `locate` (just after you have done `locate.updatedb`) in a try to find additional python interpreters in your system. Now that you are supposed to have only one python 3, it should be easier to find the one that LG got linked to (supposing this is indeed the problem).

**Comment by jmugan** [comment]
I tried `import linkgrammar` from all of the Pythons I could find using `locate bin\python` and all of them either gave me `linkgrammar not found` or the no thread error. Maybe I will wipe my whole machine and try again.

**Comment by ampli** [comment]
Please wait...
Before you do that, do the following.
Supposing your only `_clinkgrammar*.so` is in `/anaconda3/lib/python3.5/site-packages/linkgrammar/`,
put it in the import path (supposing your shell is `sh` or `bash`), and do for each python3 in your system:
```
PYTHONPATH=/anaconda3/lib/python3.5/site-packages/linkgrammar   /FULL-PATH/python
Python 3.x.y (...)
...
>>> import linkgrammar
```

**Comment by jmugan** [comment]
Cool. I'll try that and report back. Thanks!

**Comment by jmugan** [comment]
Update the DB and look for Pythons:
```
JMugan:bin jmugan$ sudo /usr/libexec/locate.updatedb

JMugan:bin jmugan$ locate bin/python
/Applications/MAMP/Library/bin/python
/Applications/MAMP/Library/bin/python-config
/Applications/MAMP/Library/bin/python2
/Applications/MAMP/Library/bin/python2-config
/Applications/MAMP/Library/bin/python2.7
/Applications/MAMP/Library/bin/python2.7-config
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/bin/python-config
/System/Library/Frameworks/Python.framework/Versions/2.7/bin/python
/System/Library/Frameworks/Python.framework/Versions/2.7/bin/python-config
/System/Library/Frameworks/Python.framework/Versions/2.7/bin/python2
/System/Library/Frameworks/Python.framework/Versions/2.7/bin/python2-config
/System/Library/Frameworks/Python.framework/Versions/2.7/bin/python2.7
/System/Library/Frameworks/Python.framework/Versions/2.7/bin/python2.7-config
/System/Library/Frameworks/Python.framework/Versions/2.7/bin/pythonw
/System/Library/Frameworks/Python.framework/Versions/2.7/bin/pythonw2
/System/Library/Frameworks/Python.framework/Versions/2.7/bin/pythonw2.7
/Users/jmugan/Envs/env1/bin/python
/Users/jmugan/Envs/env1/bin/python2
/Users/jmugan/Envs/env1/bin/python2.7
/anaconda3/bin/python
/anaconda3/bin/python.app
/anaconda3/bin/python3
/anaconda3/bin/python3-config
/anaconda3/bin/python3.5
/anaconda3/bin/python3.5-config
/anaconda3/bin/python3.5m
/anaconda3/bin/python3.5m-config
/anaconda3/bin/pythonw
/anaconda3/pkgs/python-3.5.4-h8f450c2_22/bin/python
/anaconda3/pkgs/python-3.5.4-h8f450c2_22/bin/python3
/anaconda3/pkgs/python-3.5.4-h8f450c2_22/bin/python3-config
/anaconda3/pkgs/python-3.5.4-h8f450c2_22/bin/python3.5
/anaconda3/pkgs/python-3.5.4-h8f450c2_22/bin/python3.5-config
/anaconda3/pkgs/python-3.5.4-h8f450c2_22/bin/python3.5m
/anaconda3/pkgs/python-3.5.4-h8f450c2_22/bin/python3.5m-config
/anaconda3/pkgs/python.app-2-py35he4d1c94_7/bin/python.app
/anaconda3/pkgs/python.app-2-py35he4d1c94_7/bin/pythonw
/anaconda3/pkgs/python.app-2-py36h54569d5_7/bin/python.app
/anaconda3/pkgs/python.app-2-py36h54569d5_7/bin/pythonw
/opt/homebrew-cask/Caskroom/gedit/3.2.6-3/gedit.app/Contents/Resources/bin/python
/opt/homebrew-cask/Caskroom/gedit/3.2.6-3/gedit.app/Contents/Resources/bin/python-config
/opt/homebrew-cask/Caskroom/gedit/3.2.6-3/gedit.app/Contents/Resources/bin/python2.7
/opt/homebrew-cask/Caskroom/gedit/3.2.6-3/gedit.app/Contents/Resources/bin/python2.7-config
/usr/bin/python
/usr/bin/python-config
/usr/bin/python2.7
/usr/bin/python2.7-config
/usr/bin/pythonw
/usr/bin/pythonw2.7
JMugan:bin jmugan$ 
```
then
```make clean
./configure --disable-java-bindings --enable-python-bindings=3 --enable-editline --prefix=/anaconda3/
```
shows
```
...
checking for python... /anaconda3/bin/python
checking for python version... 3.5
checking for python platform... darwin
checking for python script directory... ${prefix}/lib/python3.5/site-packages
checking for python extension module directory... ${exec_prefix}/lib/python3.5/site-packages
checking for python3.5... (cached) /anaconda3/bin/python
checking for a version of Python >= '2.1.0'... yes
checking for a version of Python >= '3.4'... yes
checking for the distutils Python package... yes
checking for Python include path... -I/anaconda3/include/python3.5m
checking for Python library path... -L/anaconda3/lib -lpython3.5m
checking for Python site-packages path... /anaconda3/lib/python3.5/site-packages
...
```
then
```
make
sudo make install
```
then
```
PYTHONPATH=/anaconda3/lib/python3.5/site-packages/linkgrammar
```
then
```
JMugan:bin jmugan$ python
Python 3.5.4 |Anaconda custom (64-bit)| (default, Nov  3 2017, 13:44:35) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import linkgrammar
Fatal Python error: PyThreadState_Get: no current thread
Abort trap: 6
JMugan:bin jmugan$ 
```
then
```
JMugan:bin jmugan$ /anaconda3/bin/python
Python 3.5.4 |Anaconda custom (64-bit)| (default, Nov  3 2017, 13:44:35) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import linkgrammar
Fatal Python error: PyThreadState_Get: no current thread
Abort trap: 6
```
then
```
JMugan:bin jmugan$ /anaconda3/bin/python3.5
Python 3.5.4 |Anaconda custom (64-bit)| (default, Nov  3 2017, 13:44:35) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import linkgrammar
Fatal Python error: PyThreadState_Get: no current thread
Abort trap: 6
JMugan:bin jmugan$ 
```
then
```
JMugan:bin jmugan$ /anaconda3/bin/python3.5m
Python 3.5.4 |Anaconda custom (64-bit)| (default, Nov  3 2017, 13:44:35) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import linkgrammar
Fatal Python error: PyThreadState_Get: no current thread
Abort trap: 6
```
then
```
JMugan:bin jmugan$ /anaconda3/bin/python3
Python 3.5.4 |Anaconda custom (64-bit)| (default, Nov  3 2017, 13:44:35) 
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import linkgrammar
Fatal Python error: PyThreadState_Get: no current thread
Abort trap: 6
```

**Comment by ampli** [comment]
The compilation stamps strongly hint that all of these are the same file.
Try to find out for sure by  `cmp FILE1 FILE2`          .
However, I cannot find additional relevant python 3 binaries in your list.

Maybe last things to try:
1. Find all python 3 `_clinkgrammar*.so`, to see if you have them in more than one directory (you shouldn't):
`locate _clinkgrammar `
If you have more than one such directory, try `python3` again after setting PYTHONPATH to it.
Also try `ls -l` on these directories if you have more than one.

2. Theoretically `_clinkgrammar*.so` could be linked to a left-over python 3 library, so you may try:
`locate lib/libpython | grep dylib`
to see if you have such unexpected libraries (use `cmp` to find out if multiple copies are identical).

**Comment by jmugan** [comment]
```
JMugan:bin jmugan$ cmp /anaconda3/bin/python /anaconda3/bin/python3.5
JMugan:bin jmugan$ cmp /anaconda3/bin/python /anaconda3/bin/python3.5m
JMugan:bin jmugan$ 
```
and
```
JMugan:bin jmugan$ locate _clinkgrammar
/Users/jmugan/bin/link-grammar-5.4.2/bindings/python/.deps/_clinkgrammar_la-lg_python_wrap.Plo
/Users/jmugan/bin/link-grammar-5.4.2/bindings/python3/.deps/_clinkgrammar_la-lg_python_wrap.Plo
/Users/jmugan/bin/link-grammar-5.4.2/bindings/python3/.libs/_clinkgrammar.5.so
/Users/jmugan/bin/link-grammar-5.4.2/bindings/python3/.libs/_clinkgrammar.a
/Users/jmugan/bin/link-grammar-5.4.2/bindings/python3/.libs/_clinkgrammar.la
/Users/jmugan/bin/link-grammar-5.4.2/bindings/python3/.libs/_clinkgrammar.lai
/Users/jmugan/bin/link-grammar-5.4.2/bindings/python3/.libs/_clinkgrammar.so
/Users/jmugan/bin/link-grammar-5.4.2/bindings/python3/.libs/_clinkgrammar_la-lg_python_wrap.o
/Users/jmugan/bin/link-grammar-5.4.2/bindings/python3/_clinkgrammar.la
/Users/jmugan/bin/link-grammar-5.4.2/bindings/python3/_clinkgrammar_la-lg_python_wrap.lo
/Users/jmugan/bin/link-grammar-5.4.2/bindings/python3/_clinkgrammar_la-lg_python_wrap.o
/anaconda3/lib/python3.5/site-packages/linkgrammar/_clinkgrammar.5.so
/anaconda3/lib/python3.5/site-packages/linkgrammar/_clinkgrammar.a
/anaconda3/lib/python3.5/site-packages/linkgrammar/_clinkgrammar.la
/anaconda3/lib/python3.5/site-packages/linkgrammar/_clinkgrammar.so
```
and
```
JMugan:bin jmugan$ locate lib/libpython | grep dylib
/Library/Frameworks/SAVI.framework/Versions/A/Frameworks/Python.framework/Versions/3.5/lib/libpython3.5.dylib
/Library/Frameworks/SAVI.framework/Versions/A/Frameworks/Python.framework/Versions/3.5/lib/libpython3.5m.dylib
/System/Library/Frameworks/Python.framework/Versions/2.7/lib/libpython2.7.dylib
/anaconda3/lib/libpython3.5m.dylib
/anaconda3/pkgs/python-3.5.4-h8f450c2_22/lib/libpython3.5m.dylib
/opt/homebrew-cask/Caskroom/gedit/3.2.6-3/gedit.app/Contents/Resources/lib/libpython2.7.dylib
/usr/lib/libpython.dylib
/usr/lib/libpython2.7.dylib
```
and
```
JMugan:bin jmugan$ cmp /Library/Frameworks/SAVI.framework/Versions/A/Frameworks/Python.framework/Versions/3.5/lib/libpython3.5.dylib /Library/Frameworks/SAVI.framework/Versions/A/Frameworks/Python.framework/Versions/3.5/lib/libpython3.5m.dylib
JMugan:bin jmugan$ 
```

**Comment by ampli** [comment]
This is a mystery for now.
I cannot think of any more checks.
I will think on that some more.

**Comment by jmugan** [comment]
Sounds good. Thanks!

**Comment by ampli** [comment]
I just updated to the latest anaconda version and was able to replicate the problem.
So it seems the problem didn't occur in older versions.

It turns out the chosen python then is in `/anaconda3/pkgs/` (check the list you sent) instead of in``/anaconda3/bin`. When I tried that python (from `/anaconda3/pkgs/`) all works fine. From some reason these python versions are different (in my system they are older).

Please try it...

I will check `configure` to find out how come it configures a python that is not even in the PATH.
Another thing to check is the location of the libraries these python versions use, and the location reported by the python that is in the path.  I already noted that the location of `libreadline` in the python binaries from the `pkgs` directory are incorrect (it notifies that on invocation...).

I will update here about my findings.

**Comment by ampli** [comment]
@jmugan ,
I finally found the reason...

You wrote:

> Using otool I get
> ```
> JMugan:~ jmugan$ otool -L /Users/jmugan/anaconda3/bin/python
> /Users/jmugan/anaconda3/bin/python:
> 	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 853.0.0)
> 	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)

Only after encountering this problem myself I noted what should have been obvious from this output: These python versions are statically linked! (Note that `otool -L` doesn't show libpython...)
When `_clinkgrammar.so` is loaded, it doesn't share its symbols with the statically linked python, and havoc happens.

The older pythons in `pkgs/bin` are still dynamically linked and it seems the older library is compatible enough for not causing problems:
```
% otool -L /Users/amir/anaconda/pkgs/python-3.5.2-0/bin/python
/Users/amir/anaconda/pkgs/python-3.5.2-0/bin/python:
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 635.21.0)
	@rpath/libpython3.5m.dylib (compatibility version 3.5.0, current version 3.5.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 159.1.0)
```

It turned it only happens starting with the recent version, as told by an anaconda developer [here](https://github.com/ContinuumIO/anaconda-issues/issues/6401):
> ... our Python 3 executables are now statically linked ...
> ... this accounts for part of the speed benefit ...
> ... we never made any guarantees about how we will link our Python ...

So if you would like to use anaconda python, you will need to use an old version (and after I updated it using `conda update` there are still such old version in the `pkgs/bin` directory that works fine with LG).

Alternatively, use Homebrew python.

Theoretically it is possible to use `pyenv` to install a usable recent python. However, a test I did once failed to import linkgrammar. But now I think I might happen to pick up a statically linked version, so I will try again.

**Comment by jmugan** [comment]
Okay, back from the code-a-thon.
Dang, even the pkgs doesn't work for me.

I created a script called `parsing.py`
```
import sys
sys.path.append('/anaconda3/lib/python3.5/site-packages/linkgrammar')
import linkgrammar
```
and when I do 
`/anaconda3/pkgs/python-3.5.4-h8f450c2_22/bin/python parsing.py`
I get the dreaded
```
Fatal Python error: PyThreadState_Get: no current thread
Abort trap: 6
```
Same story for each of `/anaconda3/pkgs/python-3.5.4-h8f450c2_22/bin/python3` and `/anaconda3/pkgs/python-3.5.4-h8f450c2_22/bin/python3.5` and `/anaconda3/pkgs/python-3.5.4-h8f450c2_22/bin/python3.5m`.

I'll try to compile linkgrammar against the system python or Homebrew.

**Comment by ampli** [comment]
Anaconda python 3.5.4 is statically linked (try `otool -L` - you will not see the python dynamic library). After a recent update at my /pkgs/ directory there is also python version is 3.5.2 which is dynamically linked).
```
% ls anaconda/pkgs/*/bin/python
anaconda/pkgs/python-3.5.2-0/bin/python
anaconda/pkgs/python-3.5.4-hb8880cc_19/bin/python
```
Try to see if you have 3.5.2 too (but I guess it may disappear on the next update).

```
% anaconda/bin/python
Python 3.5.4 |Anaconda, Inc.| (default, Oct  5 2017, 02:58:14) 
...
% % otool -L anaconda/bin/python
anaconda/bin/python:
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 853.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1197.1.1)
```
As you can see, there is no dynamic python library there.
However:
```
% anaconda/pkgs/python-3.5.2-0/bin/python
Python 3.5.2 |Continuum Analytics, Inc.| (default, Jul  2 2016, 17:52:12)
...
% otool -L anaconda/pkgs/python-3.5.2-0/bin/python
anaconda/pkgs/python-3.5.2-0/bin/python:
	/System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation (compatibility version 150.0.0, current version 635.21.0)
	@rpath/libpython3.5m.dylib (compatibility version 3.5.0, current version 3.5.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 159.1.0)
```

So unless you can see (using `otool -L` that your python is dynamically linked with the python library, there is no need even to try LG with it - it will fail.

Conclusion:
You cannot use LG with an Anaconda Python version (or any other one) if it is statically linked with **libpython**.

**Comment by jmugan** [comment]
Cool, thanks! I was able to get it working with Homebrew!
